controllers/drain_controller_test.go

```diff
@@ -316,6 +316,44 @@ var _ = Describe("Drain Controller", Ordered, func() {
 			expectNodeStateAnnotation(nodeState1, constants.Draining)
 		})
 	})
+
+	Context("With requesting device plugin restart", func() {
+		It("should remove and the device plugin label on the node with nodeState spec is empty", func() {
+			node, nodeState := createNode(ctx, "node5")
+			simulateDaemonSetAnnotation(node, constants.DevicePluginResetRequired)
+			expectNodeStateAnnotation(nodeState, constants.DrainComplete)
+			expectNodeLabel(node, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelDisabled)
+			simulateDaemonSetAnnotation(node, constants.DrainIdle)
+			expectNodeStateAnnotation(nodeState, constants.DrainIdle)
+			expectNodeLabel(node, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelDisabled)
+		})
+
+		It("should remove and add the device plugin label on the node", func() {
+			node, nodeState := createNode(ctx, "node6")
+			nodeState.Spec = sriovnetworkv1.SriovNetworkNodeStateSpec{Interfaces: sriovnetworkv1.Interfaces{
+				{
+					Name:   "test",
+					NumVfs: 10,
+					VfGroups: []sriovnetworkv1.VfGroup{
+						{
+							PolicyName:   "test",
+							ResourceName: "test",
+							VfRange:      "0-9",
+						},
+					},
+				},
+			}}
+			Expect(k8sClient.Update(ctx, nodeState)).ToNot(HaveOccurred())
+
+			simulateDaemonSetAnnotation(node, constants.DevicePluginResetRequired)
+			expectNodeStateAnnotation(nodeState, constants.DrainComplete)
+			expectNodeLabel(node, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelDisabled)
+			simulateDaemonSetAnnotation(node, constants.DrainIdle)
+			createDevicePluginPodOnNode(ctx, node.Name)
+			expectNodeStateAnnotation(nodeState, constants.DrainIdle)
+			expectNodeLabel(node, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelEnabled)
+		})
```

Comment:
While checking if I understood all the step, I coded this unit test:
```
It("should restart the device plugin on multiple nodes if requested", func() {
			simpleNodeStateSpec := sriovnetworkv1.SriovNetworkNodeStateSpec{Interfaces: sriovnetworkv1.Interfaces{
				{
					Name:   "test",
					NumVfs: 10,
					VfGroups: []sriovnetworkv1.VfGroup{
						{
							PolicyName:   "test",
							ResourceName: "test",
							VfRange:      "0-9",
						},
					},
				},
			}}

			node1, nodeState1 := createNode(ctx, "node-a")
			nodeState1.Spec = simpleNodeStateSpec
			node2, nodeState2 := createNode(ctx, "node-b")
			nodeState2.Spec = simpleNodeStateSpec


			simulateDaemonSetAnnotation(node1, constants.DevicePluginResetRequired)
			simulateDaemonSetAnnotation(node2, constants.DevicePluginResetRequired)
			
			expectNodeStateAnnotation(nodeState1, constants.DrainComplete)
			expectNodeStateAnnotation(nodeState2, constants.DrainComplete)
			expectNodeLabel(node1, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelDisabled)
			expectNodeLabel(node2, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelDisabled)

			simulateDaemonSetAnnotation(node1, constants.DrainIdle)
			simulateDaemonSetAnnotation(node2, constants.DrainIdle)

			expectNodeStateAnnotation(nodeState1, constants.DrainIdle)
			expectNodeStateAnnotation(nodeState2, constants.DrainIdle)
			expectNodeLabel(node1, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelEnabled)
			expectNodeLabel(node2, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelEnabled)
		})
```

But it fails and the device plugin is not enabled on any node. Is this test case wrong, or is there an underlying unmanaged condition?

Reply:
Done please take a look.

in your unit tests you miss to update the nodeState obect with the new spec and to also create a demo device plugin pod. I fixed it and integrate the test

Reply:
looks good, thanks!